src/hotspot/share/opto/phaseX.cpp

```diff
@@ -1974,17 +1974,26 @@ void PhaseCCP::push_load_barrier(Unique_Node_List& worklist, const BarrierSetC2*
 // AndI/L::Value() optimizes patterns similar to (v << 2) & 3 to zero if they are bitwise disjoint.
 // Add the AndI/L nodes back to the worklist to re-apply Value() in case the shift value changed.
 // Pattern: parent -> LShift (use) -> ConstraintCast* -> And
+// Pattern: parent -> LShift (use) -> ConstraintCast* -> ConvI2L -> ConstraintCast* -> And
 void PhaseCCP::push_and(Unique_Node_List& worklist, const Node* parent, const Node* use) const {
   uint use_op = use->Opcode();
   if ((use_op == Op_LShiftI || use_op == Op_LShiftL)
       && use->in(2) == parent) { // is shift value (right-hand side of LShift)
+    Node* conv = nullptr;
     auto push_and_uses_to_worklist = [&](Node* n){
       uint opc = n->Opcode();
       if (opc == Op_AndI || opc == Op_AndL) {
         push_if_not_bottom_type(worklist, n);
+      } else if (opc == Op_ConvI2L) {
+        conv = n;
       }
     };
     ConstraintCastNode::visit_uncasted_uses(use, push_and_uses_to_worklist);
+    if (conv) {
```

Comment:
What happens if there were to be multiple such `ConvI2L` nodes? Or is that impossible?

Reply:
`MulNode::AndIL_shift_and_mask_is_always_zero` only covers the case where there is one conversion. So, currently, such a case is impossible in the analysis.

Reply:
Just because a son only has one father does not mean that a father only has one son ;)

`MulNode::AndIL_shift_and_mask_is_always_zero` looks up through its inputs and only walks by one `ConvI2L` at most, on the way to the `Shift` node. But going down from a `shift`, there could be multiple `ConvI2L` nodes, and you do not know which ones would lead to the `add-shift` pattern (there could be multiple), and which ones lead to other structures.

What do you think?

Reply:
Ah, good point. So, essentially, the `if (conv)` in my fix will only check from the ConvI2L last visited by the first call to `ConstraintCastNode::visit_uncasted_uses` (which may not be the one leading to AndL).

Reply:
Exactly! Or it may be one, but there are others. Actually, it would be interesting to see if we can find an example. Maybe too much work thouch.

Reply:
Fix pushed, would be great if you could have another look @eme64. I'll try to construct an example as well where my first fix fails.